Core/include/Acts/Geometry/Detector.hpp

```diff
@@ -0,0 +1,161 @@
+  /// @note Only call this if you need shared ownership of this object.
+  ///
+  /// @return The shared pointer
+  std::shared_ptr<Detector> getSharedPtr();
+
+  /// Retrieve a @c std::shared_ptr for this surface (const version)
+  ///
+  /// @note Will error if this was not created through the @c makeShared factory
+  ///       since it needs access to the original reference. In C++14 this is
+  ///       undefined behavior, but most likely implemented as a @c bad_weak_ptr
+  ///       exception, in C++17 it is defined as that exception.
+  /// @note Only call this if you need shared ownership of this object.
+  ///
+  /// @return The shared pointer
+  std::shared_ptr<const Detector> getSharedPtr() const;
+
+  /// Non-const access to the volumes
+  ///
+  /// @return the volumes shared pointer store
+  std::vector<std::shared_ptr<DetectorVolume>>& volumePtrs();
+
+  /// Const access to sub volumes
+  ///
+  /// @return a vector to const DetectorVolume raw pointers
+  const std::vector<const DetectorVolume*>& volumes() const;
+
+  /// Update the current volume of a given navigation state
+  ///
+  /// @param gctx is the Geometry context of the call
+  /// @param nState [in, out] is the navigation state
+  ///
+  void updateDetectorVolume(const GeometryContext& gctx,
+                            NavigationState& nState) const;
+
+  /// Find a volume from a position
+  ///
+  /// @param gctx is the Geometry context of the call
+  /// @param position is the position of the call
+  ///
+  /// @note this creates internally a NavigationState object
+  ///
+  /// @return the volume pointer or nullptr (if outside)
+  const DetectorVolume* findDetectorVolume(const GeometryContext& gctx,
+                                           const Vector3& position) const;
+
+  /// Find a volume by name
+  ///
+  /// @param name with which the volume is searched for
+  ///
+  /// @return the volume pointer or nullptr (if not found)
+  const DetectorVolume* findDetectorVolume(const std::string& name) const;
+
+  /// Update the volume finder
+  ///
+  /// @param mVolumeFinder the new managed volume finder
+  void updateDetectorVolumeFinder(ManagedDetectorVolumeUpdator&& mVolumeFinder);
+
+  /// Const access to the volume finder
+  const ManagedDetectorVolumeUpdator& detectorVolumeFinder() const;
+
+  /// Return the name of the detector
+  const std::string& name() const;
+
+ private:
+  /// Name of the detector
+  std::string m_name = "Unnamed";
+
+  /// Volume store (internal/external)
+  DetectorVolume::ObjectStore<std::shared_ptr<DetectorVolume>> m_volumes;
+
+  /// A volume finder delegate
+  ManagedDetectorVolumeUpdator m_volumeFinder;
+
+  /// Name/index map
+  std::map<std::string, size_t> m_volumeNameIndex;
+};
+
+inline std::vector<std::shared_ptr<DetectorVolume>>& Detector::volumePtrs() {
+  return m_volumes.internal;
+}
+
+inline const std::vector<const DetectorVolume*>& Detector::volumes() const {
+  return m_volumes.external;
```

Comment:
Ok I see. You can sort of do that anyway by using an offset from the begin iterator, which I think should compile down to the same thing the way that the vector iterator is implemented. But ok, we can keep it like this and we (maybe I) can see if this can be achieved reasonably with iterator pairs.